proposals/ref-struct-interfaces.md

```diff
@@ -0,0 +1,195 @@
+
+To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the `class` implements the interface. 
+
+Default interface methods pose a problem for `ref struct` as there are no protections against the default implementation boxing the `this` member. 
+
+```csharp
+interface I1
+{
+    void M()
+    {
+        // Error: both of these box if I1 is implemented by a ref struct
+        I1 local1 = this;
+        object local2 = this;
+    }
+}
+
+ref struct S = I1 { }
+```
+
+To handle this a `ref struct` will be forced to implement all members of an interface, even if they have default implementations. The runtime will also be updated to throw an exception if a default interface member is called on a `ref struct` type.
+
+Detailed Notes:
+- A `ref struct` can implement an interface
+- A `ref struct` cannot participate in default interface members
+- A `ref struct` cannot be cast to interfaces it implements as that is a boxing operation
+
+### ref struct Generic Parameters
+<a name="ref-struct-generic"></a>
+The language will allow for generic parameters to opt into supporting `ref struct` as arguments by using the `allow T : ref struct` syntax: 
+
+```csharp
+T Identity<T>(T p) 
+    allow T : ref struct
+    => p;
+
+// Okay
+Span<T> local = Identity(new Span<int>(new int[10]));
+```
+
+This is similar to a `where` in that it specifies the capabilities of the generic parameter. The difference is `where` limits the set of types that can fulfill a generic parameter while the behavior defined here expands the set of types. This is effectively an anti-constraint as it removes the implicit constraint that `ref struct` cannot satisfy a generic parameter. As such this is given a new syntax, `allow`, to make that clearer.
+
+A type parameter bound by `allow T: ref struct` has all of the behaviors of a `ref struct` type: 
+
+1. Instances of it cannot be boxed
+2. Instances participate in lifetime rules like a normal `ref struct`
+3. The type parameter cannot be used in `static` fields, elements of an array, etc ...
+4. Instances can be marked with `scoped`
+
+Examples of these rules in action: 
+
+```csharp
+interface I1 { }
+I1 M1<T>(T p)
+    allow T : ref struct, I1
+{
+    // Error: cannot box potential ref struct
+    return p;
+}
+
+T M2<T>(T p)
+{
+    Span<int> span = stackalloc int[42];
+
+    // The safe-to-escape of the return is current method because one of the inputs is 
+    // current method
+    T t = M3<T>(span);
+
+    // Error: the safe-to-escape is current method.
+    return t;
+
+    // Okay
+    return default;
+    return p;
+}
+
+T M3<T>(Span<int> span)
```

Comment:
Should `M3<T>` have a `Span<T>` parameter? If so, perhaps:

```suggestion
T M2<T>(T p)
    where T : unmanaged
{
    Span<T> span = stackalloc T[42];

    // The safe-to-escape of the return is current method because one of the inputs is 
    // current method
    T t = M3<T>(span);

    // Error: the safe-to-escape is current method.
    return t;

    // Okay
    return default;
    return p;
}

T M3<T>(Span<T> span)
```

Reply:
This should be `allow T : ref struct`